backend/store/postgres/namespaces_schema.go

```diff
@@ -0,0 +1,188 @@
+SET namespace_id = namespaces.id
+FROM namespaces
+WHERE namespaces.name = entity_configs.namespace;
+
+-- Set namespace_id values in the entity_states table
+UPDATE entity_states
+SET namespace_id = namespaces.id
+FROM namespaces
+WHERE namespaces.name = entity_states.namespace;
+
+-- Add foreign key constraint to entity_configs table
+ALTER TABLE entity_configs
+ADD FOREIGN KEY (namespace_id) REFERENCES namespaces (id)
+ON DELETE CASCADE;
+
+-- Add foreign key constraint to entity_states table
+ALTER TABLE entity_states
+ADD FOREIGN KEY (namespace_id) REFERENCES namespaces (id)
+ON DELETE CASCADE;
+
+-- Update unique constraint for entity_configs table
+ALTER TABLE entity_configs
+DROP CONSTRAINT entity_config_unique;
+ALTER TABLE entity_configs
+ADD CONSTRAINT entity_config_unique UNIQUE (namespace_id, name);
+
+-- Update unique constraint for entity_states table
+ALTER TABLE entity_states
+DROP CONSTRAINT entity_state_unique;
+ALTER TABLE entity_states
+ADD CONSTRAINT entity_state_unique UNIQUE (namespace_id, name);
+
+-- Update index for entity_states table
+CREATE INDEX entity_states_name_idx ON entity_states ( namespace_id, name );
+DROP INDEX entities_name_idx;
+
+-- Drop old namespace columns
+ALTER TABLE entity_configs DROP COLUMN namespace;
+ALTER TABLE entity_states DROP COLUMN namespace;
+`
+
+const createOrUpdateNamespaceQuery = `
+-- This query creates a new namespace, or updates it if it already exists.
+--
+-- Parameters:
+-- $1: The namespace name.
+-- $2: The label selectors of the entity config.
+-- $3: The annotations of the entity config.
+--
+INSERT INTO namespaces (
+	name,
+	selectors,
+	annotations
+) VALUES ( $1, $2, $3 )
+ON CONFLICT ( name )
+DO UPDATE
+SET
+	selectors = $2,
+	annotations = $3
+`
+
+const createIfNotExistsNamespaceQuery = `
+-- This query inserts rows into the namespaces table. By design, it
+-- errors when a namespace with the same name already exists.
+--
+WITH namespace AS (
+	INSERT INTO namespaces (
+		name,
+		selectors,
+		annotations
+	) VALUES ( $1, $2, $3 )
+	RETURNING id
+)
+SELECT namespace.id FROM namespace
+`
+
+const updateIfExistsNamespaceQuery = `
+-- This query updates the namespace, but only if it exists.
+--
+WITH namespace AS (
+	SELECT id FROM namespaces
+	WHERE name = $1
+), upd AS (
+	UPDATE namespaces
+	SET
+		selectors = $2,
+		annotations = $3
+	FROM namespace
+	WHERE namespace.id = namespaces.id
+)
+SELECT * FROM namespace;
+`
+
+const getNamespaceQuery = `
+-- This query fetches a single namespace, or nothing.
+--
+SELECT
+	name,
+	selectors,
+	annotations
+FROM namespaces
+WHERE name = $1
+`
+
+const getNamespacesQuery = `
+-- This query fetches multiple namespaces.
+--
+SELECT
+	name,
+	selectors,
+	annotations
+FROM namespaces WHERE name IN (SELECT unnest($1::text[]))
+`
+
+const deleteNamespaceQuery = `
+-- This query deletes a namespace. Any related data (e.g. entity configs,
+-- entity state, etc.) will also be deleted via ON DELETE CASCADE triggers.
```

Comment:
In the etcd implementation I ended up taking special care to make sure the DeleteNamespace store method didn't orphan any resources [here](https://github.com/sensu/sensu-go/blob/6d913a110dea26ab9ede89948d8934db9a1d2c8f/backend/store/v2/etcdstore/store.go#L431-L442). Polling the team, it sounded like folks preferred to disallow deleting non-empty namespaces to the automatic cleanup approach.

Reply:
I’d like to keep the cascade triggers to prevent any orphaned relational data and then add logic to the store to prevent deletion of non-empty namespaces. This way it’s less likely to have orphaned data in the case that a namespace gets deleted outside of the store. Thoughts?

Reply:
👍  works for me. This is more of a business logic requirement than a data modeling one.

My case for not cascading deletes was that as I understand it the foreign key constraint itself is preventing orphaned data. Not sure if we get a super useful error that we could handle differently when that constraint is violated.

Reply:
Some thoughts on this.

1. It's actually useful to be able to delete entire namespaces and our customers want that.
2. It's bad to accidentally delete entire namespaces, but it's not clear that's an actual risk.
3. It's a lot more likely to orphan data when etcd is in use, especially when used in concert with postgres.

This is a bit of a usability issue. We don't offer an easy way to clear out the contents of a namespace, users have to go around and delete every resource type which is a burden.